# Exploring Ebay Car Sales Data
In this project we'll explore a dataset of used cars from ebay *Kleinanzeigen*, a classified section of the German eBay website.

It was initially scraped and uploaded to Kaggle. However, in this project, we'll only use a sample of 50,000 out of the 370,000 data points so we can run our codes rapidly. Moreover, the dataset that we have was dirtied a little to closely resemble a scraped dataset.

Here is a description of the features of our dataset:
* `dateCrawled` - When this ad was first crawled. All field-values are taken from this date.
* `name` - Name of the car.
* `seller` - Whether the seller is private or a dealer.
* `offerType` - The type of listing
* `price` - The price on the ad to sell the car.
* `abtest` - Whether the listing is included in an A/B test.
* `vehicleType` - The vehicle Type.
* `yearOfRegistration` - The year in which the car was first registered.
* `gearbox` - The transmission type.
* `powerPS` - The power of the car in PS.
* `model` - The car model name.
* `kilometer` - How many kilometers the car has driven.
* `monthOfRegistration` - The month in which the car was first registered.
* `fuelType` - What type of fuel the car uses.
* `brand` - The brand of the car.
* `notRepairedDamage` - If the car has a damage which is not yet repaired.
* `dateCreated` - The date on which the eBay listing was created.
* `nrOfPictures` - The number of pictures in the ad.
* `postalCode` - The postal code for the location of the vehicle.
* `lastSeenOnline` - When the crawler saw this ad last online.

Main goal of this project is to apply what we've learned about Pandas library specifically its usage in terms of reading, cleaning, and analyzing datasets.

---
## Data Exploration
Now let's use pandas to read and store it in a variable named `autos`.

In [1]:
# Import libraries:
import pandas as pd
import numpy as np

In [2]:
# Use pandas' read_csv and use Latin-1 to read latin characters
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [3]:
# Explore our raw dataset
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

From the information above, we can see that `vehicleType`, `gearbox`, `model`, `fuelType`, and `notRepairedDamage` have missing values. Also, out of the 20 features (or columns) 5 are numeric and 15 are object types.

We can also notice that the language used in the dataset is German (e.g. manuell, privat, kleinwagen etc.) Some values in the columns are written in strings which are supposed to be int/float types like the `price` and `odometer`. We'll fix this in the later sections.

---
## Data Cleaning
Python uses the PEP8 standard. We can recognize that the column names are in camelCase instead of snake_case. We'll fix this first to comply with the standards.

In [5]:
# Print column names
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
# Update accordingly
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']

In [7]:
# review dataset for changes
autos.head(1)

date_crawled                              name  seller offer_type  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD  privat    Angebot   

    price   abtest vehicle_type  registration_year  gearbox  power_ps   model  \
0  $5,000  control          bus               2004  manuell       158  andere   

    odometer  registration_month fuel_type    brand unrepaired_damage  \
0  150,000km                   3       lpg  peugeot              nein   

            ad_created  num_photos  postal_code            last_seen  
0  2016-03-26 00:00:00           0        79588  2016-04-06 06:45:54

Additionally, some of the column names are modified in a more descriptive way like `yearOfRegistration` to `registration_year`.

---
Let's explore more on the values of each columns using `.describe()` method for our autos dataframe.

In [8]:
# describe contents of each columns including non-numeric ones
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-12 16:06:22  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  num_photos   postal_code            last_seen  
count                 50000     50000.0  50000.000000                50000  
unique                   76         NaN           NaN                39481  
top     2016-04-03 00:00:00         NaN           NaN  2016-04-07 06:17:27  
freq                   1946         NaN           NaN                    8  
mean                    NaN         0.0  50813.627300                  NaN  
std                     NaN         0.0  25779.747957                  NaN  
min                     NaN         0.0   1067.000000                  NaN  
25%                     NaN         0.0  30451.000000                  NaN  
50%                     NaN         0.0  49577.000000                  NaN  
75%                     NaN         0.0  71540.000000                  NaN  
max                     NaN         0.0  99998.000000                  NaN

From the description above, we can see that `seller` and `offer_type` columns mostly have one value. Furthermore, as mentioned before, some column values are supposed to be in int/float type but used object type. To clearly see this, look at the `price` and `odometer` columns wherein their values have `km` and `,` which are considered to be string types. Aside from these columns, we also need to investigate the `num_photos` column whereas min and max values are `null`.

In [9]:
# Investigate num_photos
autos["num_photos"].value_counts()

0    50000
Name: num_photos, dtype: int64

In [10]:
autos["seller"].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

In [11]:
autos["offer_type"].value_counts()

Angebot    49999
Gesuch         1
Name: offer_type, dtype: int64

In our examination, `num_photos` columns are all 0. `seller` and `offer_type` columns are mostly one valued. Therefore, we'll be removing these feature columns to lessen data complexity in our analysis.

In [12]:
# Remove columns
autos.drop(["num_photos", "seller", "offer_type"], axis=1, inplace=True)

In [13]:
autos.head(1)

date_crawled                              name   price   abtest  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD  $5,000  control   

  vehicle_type  registration_year  gearbox  power_ps   model   odometer  \
0          bus               2004  manuell       158  andere  150,000km   

   registration_month fuel_type    brand unrepaired_damage  \
0                   3       lpg  peugeot              nein   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54

Let's proceed to converting `price` and `odometer` to numeric values. We'll also rename the `odometer` column into `odometer_km` to avoid losing its label.

In [14]:
def remove_special_chars(str_val):
    new_val = str_val.replace(",","")
    new_val = new_val.replace("km", "")
    new_val = new_val.replace("$","")
    return new_val

# remove special chars and convert to int
autos["price"] = autos["price"].map(remove_special_chars).astype(int)
autos["odometer"] = autos["odometer"].map(remove_special_chars).astype(int)

# rename odometer to odometer_km
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True)

In [15]:
autos.head(1)

date_crawled                              name  price   abtest  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD   5000  control   

  vehicle_type  registration_year  gearbox  power_ps   model  odometer_km  \
0          bus               2004  manuell       158  andere       150000   

   registration_month fuel_type    brand unrepaired_damage  \
0                   3       lpg  peugeot              nein   

            ad_created  postal_code            last_seen  
0  2016-03-26 00:00:00        79588  2016-04-06 06:45:54

## Exploring Odometer and Price
We might want to investigate more on `odometer_km` and `price` columns for outliers which may affect our analysis later on.

In [16]:
autos["odometer_km"].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64

We can observe that the values from this field are rounded, which might indicate that users have to select pre-set values from this field. Moreover, there are more high-kilometer cars in our sample.

In [17]:
print(autos["price"].unique().shape)
print(autos["price"].describe())
print(autos["price"].value_counts().head(10))

(2357,)
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
Name: price, dtype: int64


Similarly, we can see that `price` values are rounded but contain 2357 unique values in total. From the information above, there are 1421 cars that are priced to $0. We might also want to take a look at the higher priced cars to gauge how we can control outliers.

In [18]:
# sort descending
print(autos["price"].value_counts().sort_index(ascending=False).head(10))

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
Name: price, dtype: int64


In [19]:
# sort ascending
print(autos["price"].value_counts().sort_index(ascending=True).head(10))

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price, dtype: int64


In eBay people bid over items, and it's possible that biddings start at `$1`. On the other hand, some items (specifically second hand ones) are priced a bit too much. So we might want to limit our highest value to `$350,000` and anything above that are removed.

In [20]:
# limit our price between 1 and 350k
autos = autos[autos["price"].between(1, 350000)]
autos["price"].value_counts().sort_index(ascending=False).head(10)

350000    1
345000    1
299000    1
295000    1
265000    1
259000    1
250000    1
220000    1
198000    1
197000    1
Name: price, dtype: int64

In [21]:
# total count
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48565 entries, 0 to 49999
Data columns (total 17 columns):
date_crawled          48565 non-null object
name                  48565 non-null object
price                 48565 non-null int64
abtest                48565 non-null object
vehicle_type          43979 non-null object
registration_year     48565 non-null int64
gearbox               46222 non-null object
power_ps              48565 non-null int64
model                 46107 non-null object
odometer_km           48565 non-null int64
registration_month    48565 non-null int64
fuel_type             44535 non-null object
brand                 48565 non-null object
unrepaired_damage     39464 non-null object
ad_created            48565 non-null object
postal_code           48565 non-null int64
last_seen             48565 non-null object
dtypes: int64(6), object(11)
memory usage: 6.7+ MB


## Exploring Date columns
To make an observation about the date columns of `date_crawled`, `ad_created`, and `last_seen`, let's look at how they are formatted as dates below:

In [22]:
autos[['date_crawled','ad_created','last_seen']].head()

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [23]:
# calculate distribuation of dates (not including time)
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

In [24]:
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_values()

2016-04-07    0.001400
2016-04-06    0.003171
2016-03-18    0.012911
2016-04-05    0.013096
2016-03-06    0.014043
2016-03-13    0.015670
2016-03-05    0.025327
2016-03-24    0.029342
2016-03-16    0.029610
2016-03-27    0.031092
2016-03-25    0.031607
2016-03-17    0.031628
2016-03-31    0.031834
2016-03-10    0.032184
2016-03-26    0.032204
2016-03-23    0.032225
2016-03-11    0.032575
2016-03-22    0.032987
2016-03-09    0.033090
2016-03-08    0.033296
2016-03-30    0.033687
2016-04-01    0.033687
2016-03-29    0.034099
2016-03-15    0.034284
2016-03-19    0.034778
2016-03-28    0.034860
2016-04-02    0.035478
2016-03-07    0.036014
2016-04-04    0.036487
2016-03-14    0.036549
2016-03-12    0.036920
2016-03-21    0.037373
2016-03-20    0.037887
2016-04-03    0.038608
Name: date_crawled, dtype: float64

Crawled data are a bit uniformly distributed.

In [25]:
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64

Can be deduced also from the `last_seen` distribution that there are spikes in the values of some days. This might indicate higher sales at that time or higher reach to users.

In [26]:
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033151
2016-03-10    0.031895
2016-03-11    0.032904
2016-03-12    0.036755
2016-03-13    0.017008
2016-03-14    0.035190
2016-03-15    0.034016
2016-03-16    0.030125
2016-03-17    0.031278
2016-03-18    0.013590
2016-03-19    0.033687
2016-03-20    0.037949
2016-03-21 

Looking at `ad_created` column, ads created spans from June of 2015 until April of 2016. It can also be observed that most of the ads were created in the year or 2016.

In [27]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [28]:
autos["registration_year"].value_counts().sort_index()

1000       1
1001       1
1111       1
1800       2
1910       5
1927       1
1929       1
1931       1
1934       2
1937       4
1938       1
1939       1
1941       2
1943       1
1948       1
1950       3
1951       2
1952       1
1953       1
1954       2
1955       2
1956       4
1957       2
1958       4
1959       6
1960      23
1961       6
1962       4
1963       8
1964      12
        ... 
2000    3156
2001    2636
2002    2486
2003    2699
2004    2703
2005    2936
2006    2670
2007    2277
2008    2215
2009    2085
2010    1589
2011    1623
2012    1310
2013     803
2014     663
2015     392
2016    1220
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, Length: 95, dtype: int64

Average year of used cars that are being sold had a registration from the year 2004. Cars are invented in the late 1880s, and the listings must only be up until the year the dataset was produced which is 2016. Items with registration years that exceed between 1900s and 2016 must be removed.

In [29]:
(~autos["registration_year"].between(1900,2016)).value_counts()

False    46681
True      1884
Name: registration_year, dtype: int64

In [32]:
# percentage of cars not between 1900 and 2016
print(1884/46681)

0.040359032582849556


Only 4% would be removed from the total count; thus, it's okay to remove them.

In [30]:
# Select only items with car registration from the year 1990s till 2016
autos = autos[autos["registration_year"].between(1900, 2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
2006    0.057197
2001    0.056468
2002    0.053255
1998    0.050620
2007    0.048778
Name: registration_year, dtype: float64

In [31]:
autos["registration_year"].value_counts().head(10)

2000    3156
2005    2936
1999    2897
2004    2703
2003    2699
2006    2670
2001    2636
2002    2486
1998    2363
2007    2277
Name: registration_year, dtype: int64

We can observe that most of the cars being sold were produced from the late 1990s and 2000s.

## Exploring brands by price

In [36]:
# explore brands
print("There are", autos["brand"].nunique(), "car brands.")
autos["brand"].value_counts(normalize=True)

There are 40 car brands.


volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

It's not surprising that the top 5 brands are mostly from Germany. We can also observe that Volkswagen is the top brand listed which is twice the amount of the 2nd one (BMW). 

Since we're exploring the price per brand, it's logical to get the brands that constitute to more than 5 percent of the total count of the dataset. This would ensure a more reliable estimate of their prices.

In [40]:
brands = autos["brand"].value_counts(normalize=True)
top_brands = brands[brands > 0.05].index
top_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [43]:
# get mean price per brand
price_per_brand = {}
for brand in top_brands:
    price_per_brand[brand] = int(autos[autos["brand"] == brand]["price"].mean())

In [44]:
price_per_brand

{'audi': 9336,
 'bmw': 8332,
 'ford': 3749,
 'mercedes_benz': 8628,
 'opel': 2975,
 'volkswagen': 5402}

We can infer from the results that `Audi`, `BMW`, and `Mercedez Benz` have the highest resale prices, and it's most likely because they are top tier European cars. On the other hand, `Volkswagen`, `Ford`, and `Opel` prices revolve around 5k to 3k wherein we could see a gap between first-rate and second-rate cars. In my opinion, second-rate cars are more reasonable and affordable to buy for a second hand car.

Apart from prices, it is also essential to consider their mileages (or odometer) because they greatly affect the price. So let's create a table consisting of mean price and mean mileage per brand to investigate if there is a correlation between the two.

In [47]:
# create a series from mean_price dictionary
brand_prices = pd.Series(price_per_brand)
# create a dataframe for top brands
df = pd.DataFrame(brand_prices, columns=["mean_price"])

In [48]:
df

mean_price
audi                 9336
bmw                  8332
ford                 3749
mercedes_benz        8628
opel                 2975
volkswagen           5402

In [51]:
# create mean mileage per brand
km_per_brand = {}
for brand in top_brands:
    km_per_brand[brand] = int(autos[autos["brand"] == brand]["odometer_km"].mean())

# transform into series and add to dataframe
brand_km = pd.Series(km_per_brand)
df["mean_km"] = brand_km

In [57]:
df.sort_values(by="mean_price", ascending=False)

mean_price  mean_km
audi                 9336   129157
mercedes_benz        8628   130788
bmw                  8332   132572
volkswagen           5402   128707
ford                 3749   124266
opel                 2975   129310

In [56]:
df.corr()

mean_price   mean_km
mean_price    1.000000  0.612335
mean_km       0.612335  1.000000

In [58]:
# create mean registration year per brand
year_per_brand = {}
for brand in top_brands:
    year_per_brand[brand] = int(autos[autos["brand"] == brand]["registration_year"].mean())

# transform into series and add to dataframe
brand_year = pd.Series(year_per_brand)
df["mean_year"] = brand_year

In [60]:
df.sort_values(by="mean_price", ascending=False)

mean_price  mean_km  mean_year
audi                 9336   129157       2004
mercedes_benz        8628   130788       2002
bmw                  8332   132572       2003
volkswagen           5402   128707       2002
ford                 3749   124266       2002
opel                 2975   129310       2002

From the analysis, we can see that there is a positive correlation but it doesn't mean causation. It was only a coincidence that top tier cars (e.g. mercedez, bmw, and audi) are more used than low end cars which had lower mileage/odometer. In other words, it doesn't make sense to say that as the mileage increases, the price also increase. Because in general, lower mileage car must cost a bit more than higher mileage car.

---
## Conclusion
In this project, we have maximized the usage of Pandas library in cleaning and exploring the dataset of German used car listings in eBay. Our analysis indicated that Volkswagen is the typical car of an average German. I may even consider it as the best second-car to get for its reasonable price and popularity.


Next steps we may consider:
* Data cleaning next steps:
    * Identify categorical data that uses german words, translate them and map the values to their english counterparts
    * Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
    * See if there are particular keywords in the name column that you can extract as new columns
* Analysis next steps:
    * Find the most common brand/model combinations
    * Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the milage.
    * How much cheaper are cars with damage than their non-damaged counterparts?